psci/PSCi.hs

```diff
@@ -380,17 +381,50 @@ printModuleSignatures :: P.ModuleName -> P.Environment -> PSCI ()
+                    (outputStr   . unlines . map showDataConstructor . sortBy compareDatatypes . map (findDataConstructor dataConstructorsEnv)) moduleDataConstructors
+                    (outputStrLn . unlines . sort . map (showNameType . findNameType namesEnv)) moduleNamesIdent
+
+  where findNameType :: M.Map (P.ModuleName, P.Ident) (P.Type, P.NameKind, P.NameVisibility) -> (P.ModuleName, P.Ident) -> (P.Ident, Maybe (P.Type, P.NameKind, P.NameVisibility))
+        findNameType envNames m@(_, mIdent) = (mIdent, M.lookup m envNames)
+        showNameType :: (P.Ident, Maybe (P.Type, P.NameKind, P.NameVisibility)) -> String
+        showNameType (mIdent, Just (mType, _, _)) = P.showIdent mIdent ++ " :: " ++ P.prettyPrintType mType
+        showNameType _ = error "The impossible happened in printModuleSignatures."
+        findDataConstructor :: M.Map (P.Qualified P.ProperName) (P.DataDeclType, P.ProperName, P.Type, [P.Ident]) -> P.Qualified P.ProperName -> (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident]))
+        findDataConstructor envDataCons name = (name, M.lookup name envDataCons)
+        compareDatatypes :: (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident])) -> (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident])) -> Ordering
+        compareDatatypes (_, Just (_, dtName1, _, _)) (_, Just (_, dtName2, _, _)) = dtName1 `compare` dtName2
+        compareDatatypes _ _ = error "The impossible happened in printModuleSignatures."
+        showDataConstructor :: (P.Qualified P.ProperName, Maybe (P.DataDeclType, P.ProperName, P.Type, [P.Ident])) -> String
+        showDataConstructor (P.Qualified _ name, Just (_, _, dtType, _)) = P.runProperName name ++ " :: " ++ P.prettyPrintType dtType
+        showDataConstructor _ = error "The impossible happened in printModuleSignatures."
+        findType :: M.Map (P.Qualified P.ProperName) (P.Kind, P.TypeKind) -> P.Qualified P.ProperName -> (P.Qualified P.ProperName, Maybe (P.Kind, P.TypeKind))
+        findType envTypes name = (name, M.lookup name envTypes)
+        showType :: M.Map (P.Qualified P.ProperName) ([(String, Maybe P.Kind)], P.Type) -> (P.Qualified P.ProperName, Maybe (P.Kind, P.TypeKind)) -> Maybe String
+        showType typeSynonymsEnv (n@(P.Qualified _ name), typ) =
+          case (typ, M.lookup n typeSynonymsEnv) of
+            (Just (_, P.TypeSynonym), Just (_, dtType)) ->
+                Just ("type " ++ P.runProperName name ++ " = " ++ P.prettyPrintType dtType)
+
+            (Just (_, P.DataType _ pt), _) ->
+                Just ("data " ++ P.runProperName name ++ if null pt then "" else " = " ++ printCons pt)
+
+            _ ->
+              Nothing
+
+          where printCons pt = intercalate " | " (map (\(cons,idents) -> unwords (P.runProperName cons : map prettyPrintType idents)) pt)
```

Comment:
There might be some duplication with `psc-docs` here. Perhaps this can be made into common code?